controllers/datasciencecluster/datasciencecluster_controller.go

```diff
@@ -18,45 +18,202 @@ package datasciencecluster
 
 //+kubebuilder:rbac:groups=datasciencecluster.opendatahub.io,resources=datascienceclusters,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=datasciencecluster.opendatahub.io,resources=datascienceclusters/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=datasciencecluster.opendatahub.io,resources=datascienceclusters/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=services;namespaces;serviceaccounts;secrets;configmaps;operatorgroups,verbs=get;list;watch;create;update;patch
+//+kubebuilder:rbac:groups=addons.managed.openshift.io,resources=addons,verbs=get;list
+//+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=rolebindings;roles;clusterrolebindings;clusterroles,verbs=get;list;watch;create;update;patch
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the DataScienceCluster object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
 func (r *DataScienceClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	r.Log.Info("Reconciling DataScienceCluster resources", "Request.Namespace", req.Namespace, "Request.Name", req.Name)
+
+	instance := &dsc.DataScienceCluster{}
+	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	if err != nil && apierrs.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Start reconciling
+	if instance.Status.Conditions == nil {
+		reason := status.ReconcileInit
+		message := "Initializing DataScienceCluster resource"
+		status.SetProgressingCondition(&instance.Status.Conditions, reason, message)
+		instance.Status.Phase = status.PhaseProgressing
+		err = r.Client.Status().Update(ctx, instance)
+		if err != nil {
+			r.Log.Error(err, "Failed to add conditions to status of DataScienceCluster resource.", "DataScienceCluster", req.Namespace, "Request.Name", req.Name)
+			// r.Recorder.Eventf(instance, corev1.EventTypeWarning, "DataScienceClusterReconcileError",
+			// 	"Failed to add conditions to DataScienceCluster instance %s", instance.Name)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Check if the DataScienceCluster was deleted
+	deleted := instance.GetDeletionTimestamp() != nil
+	finalizers := sets.NewString(instance.GetFinalizers()...)
+	if deleted {
+		if !finalizers.Has(finalizer) {
+			r.Log.Info("DataScienceCluster instance has been deleted.", "instance", instance.Name)
+			return ctrl.Result{}, nil
+		}
+		r.Log.Info("Deleting DataScienceCluster instance", "instance", instance.Name)
+
+		// 	// TODO: implement delete logic
```

Comment:
can we consolidate this with the reconciliation logic below? this whole block almost feels redundant now.